### PR TITLE
add fromList tests

### PR DESCRIPTION
**Pull request by int-e**
This is adding some test cases that were previously part of #658.

Note that one test is currently failing because `Data.IntMap.fromAscList` is a bit lazier than expected.

### CODE REVIEW

File: containers-tests/tests/Utils/IsUnit.hs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+same :: a -> a -> Bool
+same x y = case reallyUnsafePtrEquality# x y of
+    0# -> False
+    _  -> True
+
+-- | Check whether the argument is a fully evaluated unit `()`.
+--
+-- Uses `reallyUnsafePtrEquality#`.
+isUnit :: () -> Bool
+isUnit = same ()
```
Comment (int-e) [inline containers-tests/tests/Utils/IsUnit.hs]: I'm wondering whether I'm reinventing a wheel here. This seems closely related to what ChasingBottoms is doing, but I didn't find anything like this there.

**Comment by treeowl** [comment]
Do you consider the test or the code correct? Or neither?

**Comment by int-e** [comment]
I believe the test is correct (it will pass once #658 is merged).

**Comment by treeowl** [comment]
Can you rebase please?

**Comment by int-e** [comment]
> Can you rebase please?

Done. (I don't know whether the force-push itself causes a notification.)

File: containers-tests/tests/Utils/IsUnit.hs (hunk 1 above)
Comment (treeowl) [reply containers-tests/tests/Utils/IsUnit.hs]: This smells funny/fragile to me. I would generally feel more comfortable with an approach that uses `unsafePerformIO` to find out when/how many times things are evaluated.

Comment (treeowl) [reply containers-tests/tests/Utils/IsUnit.hs]: Oh, but of course we can't always do that .... Hrmmm. Could you at least rename `same` to match our internal pointer equality function?

Comment (int-e) [reply containers-tests/tests/Utils/IsUnit.hs]: Oh I wasn't aware of `ptrEq`, or I would have used that; will change.

I should probably make the tests conditional on __GLASGOW_HASKELL__ being defined... and  maybe make isUnit tristate? (Yes | No | Maybe; the last indicating that the result is unknown.)

**Comment by treeowl** [comment]
The test suite is already pretty GHC-specific, so I wouldn't worry about
that. Tristate ... not sure. Depends what you're trying to check. Add more
comments?

File: containers-tests/tests/Utils/IsUnit.hs (hunk 1 above)
Comment (int-e) [reply containers-tests/tests/Utils/IsUnit.hs]: > Oh I wasn't aware of `ptrEq`, or I would have used that; will change.

Ah it's not so easy because the module is hidden, and we probably want to keep the library parts of `containers-tests` and `containers` the same.

> I should probably make the tests conditional on `__GLASGOW_HASKELL__` being defined... and maybe make isUnit tristate? (Yes | No | Maybe; the last indicating that the result is unknown.)

I will go with `Maybe Bool`.

Comment (int-e) [reply containers-tests/tests/Utils/IsUnit.hs]: Given your later comment, I stuck with `Bool`. I did add some logic to deal with non-ghc implementations though.

**Comment by sjakobi** [comment]
@int-e, @treeowl, what's the status here?

**Comment by int-e** [comment]
Nothing new from my side... the branch rebased cleanly on current master, and the tests currently pass (for me).

File: containers-tests/tests/intmap-strictness.hs [hunk 2]
```diff
@@ -76,6 +82,70 @@ pInsertLookupWithKeyValueStrict f k v m
                      not (isBottom $ M.insertLookupWithKey (const3 1) k bottom m)
     | otherwise    = isBottom $ M.insertLookupWithKey (apply3 f) k bottom m
 
+pFromAscListLazy :: [Int] -> Bool
+pFromAscListLazy ks = not . isBottom $ M.fromAscList elems
+  where
+    ks' = map head . group . sort $ ks
```
Comment (sjakobi) [inline containers-tests/tests/intmap-strictness.hs:88]: Why not just use `Data.Containers.ListUtils.nubInt`? And why do the elements have to be unique at all? We're not testing `fromDistinctAscList` here.

File: containers-tests/tests/intmap-strictness.hs [hunk 3]
```diff
@@ -76,6 +82,70 @@ pInsertLookupWithKeyValueStrict f k v m
...
+    | otherwise = isBottom $ M.fromList elems
+  where
+    ks' = map head . group . sort $ ks
+    elems = [(k, v) | k <- ks', v <- [undefined, undefined, ()]]
+
+------------------------------------------------------------------------
+-- check for extra thunks
+
+tExtraThunksM :: Test
+tExtraThunksM = testGroup "IntMap.Strict - extra thunks" $
```
Comment (sjakobi) [inline containers-tests/tests/intmap-strictness.hs:103]: What are extra thunks? Maybe add a brief explanation and/or a reference to a definition or discussion.

File: containers-tests/tests/intmap-strictness.hs [hunk 4]
```diff
@@ -76,6 +82,70 @@ pInsertLookupWithKeyValueStrict f k v m
...
+    check :: TestName -> Bool -> IntMap () -> Test
+    check n e m = testCase n $ case M.lookup 42 m of
+        Just v -> assertBool msg (e == isUnit v)
+        _      -> assertString "key not found"
+      where
+        msg | e         = "too lazy -- expected fully evaluated ()"
+            | otherwise = "too strict -- expected a thunk"
+
+tExtraThunksL :: Test
+tExtraThunksL = testGroup "IntMap - extra thunks" $
```
Comment (sjakobi) [inline containers-tests/tests/intmap-strictness.hs:126]:
```suggestion
tExtraThunksL = testGroup "IntMap.Lazy - extra thunks" $
```

File: containers-tests/tests/intmap-strictness.hs [hunk 5]
```diff
@@ -76,6 +82,70 @@ pInsertLookupWithKeyValueStrict f k v m
...
+    ]
+  where
+    m0 = M.singleton 42 ()
+    check :: TestName -> Bool -> IntMap () -> Test
+    check n e m = testCase n $ case M.lookup 42 m of
+        Just v -> assertBool msg (e == isUnit v)
+        _      -> assertString "key not found"
+      where
+        msg | e         = "too lazy -- expected fully evaluated ()"
+            | otherwise = "too strict -- expected a thunk"
```
Comment (sjakobi) [inline containers-tests/tests/intmap-strictness.hs:123]: Looks like we never use `check ... False` here, so this could possibly be simplified?!

File: containers-tests/tests/intmap-strictness.hs (hunk 2 above)
Comment (int-e) [reply containers-tests/tests/intmap-strictness.hs]: Duplicate keys would actually invalidate this check; `fromAscList [(0,undefined),(0,()),(0,undefined),(0,())]` is bottom. (Not saying that this is how it should be... but that's how it is and has been for a long time.) 

I was not actively aware of `nubInt`, somehow.

Anyway I've addressed this and the other comments as well; I found a few minor discrepancies on top (like `Map` tests having `IntMap` in their name...)

File: containers-tests/tests/intmap-strictness.hs [hunk 6]
```diff
@@ -76,6 +82,76 @@ pInsertLookupWithKeyValueStrict f k v m
                      not (isBottom $ M.insertLookupWithKey (const3 1) k bottom m)
     | otherwise    = isBottom $ M.insertLookupWithKey (apply3 f) k bottom m
 
+------------------------------------------------------------------------
+-- test a corner case of fromAscList: if the list contains duplicate
+-- keys, then (only) the first of the given values is not evaluated.
```
Comment (sjakobi) [inline containers-tests/tests/intmap-strictness.hs:87]: That seems like a surprising property to me. I think https://github.com/haskell/containers/issues/473 tracks what we could about it. Could you add a reference to that issue?

Comment (int-e) [reply containers-tests/tests/intmap-strictness.hs]: done.

**Comment by sjakobi** [comment]
> @treeowl, please review!

Ping @treeowl! :)
